### PR TITLE
Update sqlparser to 0.24.0 

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Re https://github.com/apache/arrow-datafusion/issues/3606

(based off dependabot https://github.com/apache/arrow-datafusion/pull/3654)

 # Rationale for this change
Upgrade to get support for explain plan format (e.g. `explain graphviz`)

# What changes are included in this PR?
1. Upgrade dependency
2. Update code to use new code

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -152,6 +152,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 verbose,
                 statement,
                 analyze,
+                format: _,
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:155]: here is the support for explain plan format

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -1720,22 +1721,22 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 ],
             }),
 
-            SQLExpr::Value(Value::Interval {
+            SQLExpr::Array(arr) => self.sql_array_literal(arr.elem, schema),
+
+            SQLExpr::Interval {
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:1726]: Intervals are no longer literals (`Value`) but can have exprs so the AST was updated to match in https://github.com/sqlparser-rs/sqlparser-rs/pull/609

**Review by xudong963 — APPROVED** [review]

**Comment by xudong963** [comment]
https://github.com/apache/arrow-datafusion/actions/runs/3164334070/jobs/5152557761

Other PRs also encountered the issue :/

**Comment by alamb** [comment]
>  https://github.com/apache/arrow-datafusion/actions/runs/3164334070/jobs/5152557761

Other PRs also encountered the issue :/

🤔  maybe we have hit the space limit on the builders

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/3678 for the CI failure

**Pull request merged**